activerecord/lib/arel/nodes/annotation.rb

```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
+module Arel # :nodoc: all
+  module Nodes
+    class Annotation < Arel::Nodes::Node
```

Comment:
Optimizer hints and annotation nodes are both serialized as comments in SQL, thanks to the SQL visitor, but as Arel nodes, does OptimizerHints have an _is-a_ relationship with Annotation? Consider a `Comment` mixin (or superclass?) that clarifies that their relationship is implementation only.

Furthermore, should comment sanitization be the node's responsibility, or the SQL visitor's?

/cc @kamipo

Reply:
I think in my mental model, `Annotation` was the superclass, representing a non-executable (hand-wavey term) piece of metadata added to the query. Maybe `Annotation`, which handles comment sanitization, is the mixin and `OptimizerHints` and `Comment` are the concrete classes.

> Furthermore, should comment sanitization be the node's responsibility, or the SQL visitor's?

I do think SQL comment sanitization should be done at the node level.

At the node level it happens once, up front, at node creation time, rather than having to be repeated through multiple visitor methods. If a relation is used to generate multiple queries, the sanitization cost is paid once also.

All types of annotations have to care about SQL comment injection, so I think it makes sense to centralize it.


From https://github.com/rails/rails/pull/35615#pullrequestreview-215377223:
> Does this need different sanitization than `/* */` removal? Consider the hint `) …other SQL clauses… OPTION(`.

That would result in a query error, as multiple `OPTION` clauses in a statement is invalid. (Doc [here](https://docs.microsoft.com/en-us/sql/t-sql/queries/option-clause-transact-sql?view=sql-server-2017).)

I think the risk profile there is much lower and probably there's no need to sanitize it. (Sanitization would probably just result in a differently invalid query.) But if it did need to be escaped, that seems like something to handle in the database-specific visitor.